compiler/src/iree/compiler/Codegen/Common/FastMathDeviceLib.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/FastMathDeviceLib.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/PatternMatch.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+// Pattern to lower math.erf to its device lib implementation (from erfF.cl)
+struct ErfDeviceLibPattern : public OpRewritePattern<math::ErfOp> {
+  using OpRewritePattern<math::ErfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(math::ErfOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.getOperand();
+    Type resultType = op.getType();
+    if(resultType.isF16()) {
+        input = rewriter.create<arith::FPToSIOp>(loc, rewriter.getF32Type(), input);
```

Comment:
I forget how @bjacob set this up, but promotion should be an orthogonal decision to using the device lib. I would stick to just f32 support here and we can add a separate pattern set for promotion of math ops.